docs/rfcs/0015-pipecd-plugin-arch-meta.md

```diff
@@ -0,0 +1,143 @@
+kind: Application
+metadata:
+  labels:
+    kind: KUBERNETES # <- like this
+spec:
+  name: myApp
+```
+
+The control plane will be updated so that it can accept platform provider from pipedv0 and deployTargets from pipedv1.
+
+### The protocol
+
+The protocol means to protocol used to connect between piped and its plugins.
+As at the starting point, we have 2 options
+- WASM
+- gRPC
+
+For less study maintaining cost, we choose gRPC. Also, other components connections in PipeCD system used gRPC as well.
+
+![](assets/0015-piped-protocol.png)
+
+Boostrap flow:
+- Start piped (as entrypoint of piped pod container for example)
+- Piped loads plugins' configuration
+- Piped pull plugins from registry (if it's remote URI specified)
+- Piped starts plugins with plugins configuration passed at this point
+- Plugins start its grpc servers
+- Plugins ping and send status ready to piped plugin management grpc server
+-> ready to be used at this point
+
+In case of failure:
+- Piped failured: Restart containers / taskset as piped is the entrypoint
+- Plugins failured:
+  - Ping failed -> piped will be notified that plugins are not working
+  - Based on piped logic, will restart the failured plugins (plugin management grpc logic)
+
+### The configuration
+
+The configuration which up-to-now placed in piped and application (kind specific) manifests will be rearrangement as below
+
+![](assets/0015-config-context.png)
+
+We need futher discussion on this but currently, we agreed to separate configuration as piped context and plugin context as above.
+
+### The interfaces
+
+The Piped agent has serveral main features, which are
+
+- Plan and execute deployment
+- Calculate the plan preview result
+- Calculate the drift between live and "git source" manifests
+- Fetch and build the state of the live manifest
+- Insight feature
+
+Across these features, we have to support 5 kinds of platforms (up to now), and each deployment we have serveral kind of stages, which are
+
+- K8s stages
+- Cloud run stages
+- Terraform stages
+- ECS stages
+- Lambda stages
+- Wait stage
+- Wait approval stage
+- Analysis stage
+- Script run stage
+- Custom sync stage (deprecated)
+
+Based on the requirement, we would create plugins based on 3 big interfaces
+
+- Deployment interface: response to how deployment should be done (include: plan, execute deployment tasks)
+- LiveState interface: response to fetch and build state of live manifest
+- Drift interface: response to calculate the drift between 2 given manifests (live - git or running - target)
+
+Plugin can choose how and which interface it implement such as
+
+- K8s plugin can implement all 3 interfaces
+- Wait approval plugin can implement only Deployment interface (since it has nothing to deal with livestate and drift features)
+
+Also, we may need support interface which implemented in piped side (optional) for tasks such as: Decrypt piped secret.
+
+### The running flow
+
+# Unresolved questions
+
```

Comment:
How about adding "how to manage plugin libraries"?

There are some funcs used across plugins.
e.g. diff, launching a grpc server(?)